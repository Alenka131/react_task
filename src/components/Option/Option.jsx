import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import useStyles from './styles';

const Option = ({
  handleClick,
  text,
  selected,
  testId,
}) => {
  const classes = useStyles();

  const optionClassName = clsx({
    [classes.option]: true,
    [classes.selected]: selected,
  });

  return (
    <li className={optionClassName} data-testid={testId}>
      <button
        type="button"
        className={classes.button}
        onClick={handleClick}
      >
        {text}
      </button>
    </li>
  );
};

Option.defaultProps = {
  selected: false,
};

Option.propTypes = {
  handleClick: PropTypes.func.isRequired,
  testId: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  selected: PropTypes.bool,
};

export default Option;
