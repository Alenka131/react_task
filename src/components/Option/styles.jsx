import { makeStyles } from '@material-ui/core/styles';
import COLORS from 'shared/colors';

const UNDERLINE_STYLES = {
  content: "''",
  position: 'absolute',
  bottom: 0,
  height: 2,
};

export default makeStyles({
  button: {
    backgroundColor: COLORS.TRANSPARENT,
    borderColor: COLORS.TRANSPARENT,
  },
  option: {
    color: COLORS.WHITE,
    display: 'inline-block',
    listStyleType: 'none',
    paddingBottom: 2,
    position: 'relative',
    '&::before': {
      ...UNDERLINE_STYLES,
      left: '50%',
      transition: 'all 0.3s ease-out',
      width: 0,
    },
    '&:focus, &:hover, &:active': {
      '&::before': {
        backgroundColor: COLORS.WHITE,
        left: 0,
        width: '100%',
      },
    },
    '& ~ &': {
      marginLeft: 32,
    },
    '& > $button': {
      color: 'inherit',
      cursor: 'pointer',
      fontWeight: 'inherit',
      margin: -16,
      padding: 16,
    },
  },
  selected: {
    '&::before': {
      ...UNDERLINE_STYLES,
      backgroundColor: COLORS.WHITE,
      left: 0,
      width: '100%',
    },
    '& > $button': {
      cursor: 'default',
    },
  },
});
