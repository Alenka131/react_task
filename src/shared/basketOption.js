export const DEFAULT_BASKET_SUMMARY = 0;
export const ADD_REMOVE_ITEM_UNIT = 1;
export const MIN_ITEM_QUANTITY = 0;
