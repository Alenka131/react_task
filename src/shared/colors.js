export default {
  TRANSPARENT: 'transparent',
  WHITE: '#ffffff',
};
