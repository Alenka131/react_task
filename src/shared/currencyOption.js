export const PLN = 'PLN';
export const USD = 'USD';
export const DEFAULT_CURRENCY = PLN;
export const OPTIONS = [PLN, USD];
