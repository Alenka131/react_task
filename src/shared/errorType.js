export const INCORRECT_BODY = 'INCORRECT_BODY';
export const INCORRECT_TYPE = 'INCORRECT_TYPE';
export const MISSING_PROPERTY = 'MISSING_PROPERTY';
export const NOT_FOUND = 'NOT_FOUND';
export const INCORRECT_QUANTITY = 'INCORRECT_QUANTITY';
