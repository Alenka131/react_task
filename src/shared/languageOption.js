export const PL = 'pl';
export const EN = 'en';
export const DEFAULT_LANGUAGE = PL;
export const OPTIONS = [PL, EN];
export const PRICE_WITH_DOT_FORMAT = [EN];
export const PRICE_WITH_COMMA_FORMAT = [PL];
