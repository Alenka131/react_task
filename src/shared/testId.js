export const BASKET_PRICE_TEST_ID = 'basket_price';
export const CHANGE_CURRENCY_TEST_ID = 'change_currency';
export const CHANGE_LANGUAGE_TEST_ID = 'change_language';
export const CURRENCY_OPTION_TEST_ID = 'currency_option';
export const LANGUAGE_OPTION_TEST_ID = 'language_option';
export const LOADING_TEST_ID = 'loading';
export const MENU_TEST_ID = 'menu';
export const PRODUCT_LIST_TEST_ID = 'product_list';
export const PRODUCT_NAME_TEST_ID = 'product_name';
export const PRODUCT_PRICE_TEST_ID = 'product_price';
export const PRODUCT_TEST_ID = 'product';
export const PRODUCT_QUANTITY_TEST_ID = 'product_quantity';
export const ADD_ITEM_BUTTON_TEST_ID = 'add_item_button';
export const REMOVE_ITEM_BUTTON_TEST_ID = 'remove_item_button';
export const OPTION_DEFAULT_TEST_ID = 'option';
