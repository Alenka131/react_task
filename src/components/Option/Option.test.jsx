import React from 'react';
import { shallow } from 'enzyme';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { OPTION_DEFAULT_TEST_ID as testId } from 'shared/testId';
import Option from '.';

afterEach(cleanup);

describe('<Option />', () => {
  it('should defined', () => {
    expect(Option).not.toBeUndefined;
  });

  it('should render button in item of list', () => {
    const mockCallBack = jest.fn();
    const text = 'Lorem ipsum';

    const { getByTestId } = render(
      <Option
        handleClick={mockCallBack}
        text={text}
        testId={testId}
      />,
    );
    const option = getByTestId(testId);
    expect(option.tagName.toLowerCase()).toEqual('li');
    const button = option.querySelectorAll('button');
    expect(button).toHaveLength(1);
  });

  it('should render text property', () => {
    const mockCallBack = jest.fn();
    const text = 'Lorem ipsum';

    const { getByTestId } = render(
      <Option
        handleClick={mockCallBack}
        text={text}
        testId={testId}
      />,
    );
    const option = getByTestId(testId);
    const button = option.querySelector('button');
    fireEvent.click(button);
    expect(button.innerHTML).toEqual(text);
  });

  it('should render with additional className if has selected property', () => {
    const mockCallBack = jest.fn();
    const text = 'Lorem ipsum';

    const { getByTestId } = render(
      <Option
        handleClick={mockCallBack}
        text={text}
        testId={testId}
        selected
      />,
    );

    const option = getByTestId(testId);
    const button = option.querySelector('button');
    expect(button).toHaveStyle('cursor: default');
  });

  it('simulates click events', () => {
    const mockCallBack = jest.fn();
    const text = 'Lorem ipsum';

    const { getByTestId } = render(
      <Option
        handleClick={mockCallBack}
        text={text}
        testId={testId}
      />,
    );
    const option = getByTestId(testId);
    const button = option.querySelector('button');
    fireEvent.click(button);
    expect(mockCallBack).toHaveBeenCalledTimes(1);
  });

  it('should renders correctly', () => {
    const mockCallBack = jest.fn();
    const text = 'Lorem ipsum';

    const wrapper = shallow(
      <Option
        handleClick={mockCallBack}
        text={text}
        testId={testId}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders correctly, with selected property', () => {
    const mockCallBack = jest.fn();
    const text = 'Lorem ipsum';

    const wrapper = shallow(
      <Option
        handleClick={mockCallBack}
        text={text}
        testId={testId}
        selected
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
