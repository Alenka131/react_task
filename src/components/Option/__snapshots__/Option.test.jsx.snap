// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Option /> should renders correctly 1`] = `
<li
  className="makeStyles-option-14"
  data-testid="option"
>
  <button
    className="makeStyles-button-13"
    onClick={[MockFunction]}
    type="button"
  >
    Lorem ipsum
  </button>
</li>
`;

exports[`<Option /> should renders correctly, with selected property 1`] = `
<li
  className="makeStyles-option-14 makeStyles-selected-15"
  data-testid="option"
>
  <button
    className="makeStyles-button-13"
    onClick={[MockFunction]}
    type="button"
  >
    Lorem ipsum
  </button>
</li>
`;
